### PR TITLE
Adds controller-gen as a pinned tool

### PR DESCRIPTION
**Pull request by chuckha**
Signed-off-by: Chuck Ha <chuckh@vmware.com>

**What this PR does / why we need it**:
This PR starts to make CAPA look more like the other providers by pinning controller-gen in the tools repository.

We still use the vendor version but the overall goal is to replace that with the recommended go path for pinning dev/test tooling. 

We are not using bazel to run controller-gen so this is not a major departure from existing tooling.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #968 

**Special notes for your reviewer**:
This code is not yet being used but will be used in existing makefile targets.

**Release note**:
```release-note
NONE
```
/assign @detiber @ncdc

### CODE REVIEW

**Comment by detiber** [comment]
@chuckha should there also be related Makefile changes?

**Comment by chuckha** [comment]
@detiber not yet, that's coming. This is just adding the tooling. (also update for fixed boilerplate)

**Comment by vincepri** [comment]
/hold

How is this different from https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/hack/tools.go? Is there any benefit from one vs the other?

**Comment by ncdc** [comment]
@vincepri I chatted upstream and people suggested using a separate go "sub"module for tools. We should convert CAPA to this approach and migrate what's in hack/tools.go here.

**Comment by chuckha** [comment]
I can move over all the tools.go imports into this submodule

**Comment by vincepri** [comment]
I'm fine either way, I wonder if any other project is following the sub-module approach. AFAIK controller-runtime and controller-tools only have `hack/tools.go`. When I was working on CAPI a while back I tried both approaches, in the end I preferred to keep everything in one `go.mod` to easily track dependencies and their replacements (if necessary). 

If we use the sub-module approach, is there going to be a case where we might have to deal with  version compatibility issues?

**Comment by ncdc** [comment]
Maybe? Let's cross that bridge if/when we need to?

**Comment by vincepri** [comment]
Sounds good, @chuckha can you also do the same for the other imports? We should do the same for CAPI as well to standardize across all repos.

Feel free to unhold.

**Comment by chuckha** [comment]
I would strongly recommend going through this commit by commit if you're willing to review.

Contentious points:

specifying clusterctl//capi dependency version in two places.

File: config/crd/bases/_.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  names:
+    kind: ""
+    plural: ""
+  scope: ""
+status:
+  acceptedNames:
+    kind: ""
+    plural: ""
+  conditions: null
+  storedVersions: null
```
Comment (detiber) [inline config/crd/bases/_.yaml]: Is this file correct?

File: WORKSPACE [hunk 2]
```diff
@@ -109,12 +109,6 @@ go_repository(
     tag = "v1.3.1",
 )
 
-go_repository(
-    name = "io_k8s_sigs_kustomize",
-    importpath = "sigs.k8s.io/kustomize",
-    tag = "v1.0.11",
-)
-
```
Comment (detiber) [inline WORKSPACE]: Are all the bazel targets that were using this removed?

File: config/crd/bases/_.yaml (hunk 1 above)
Comment (chuckha) [reply config/crd/bases/_.yaml]: how did that get in there. No, this file should not be in here.

File: WORKSPACE (hunk 2 above)
Comment (chuckha) [reply WORKSPACE]: No, I need to fix that up but I've run out of time for the day

**Comment by chuckha** [comment]
/close

I'm going for a different approach, this was too much at once

**Pull request closed without merging**